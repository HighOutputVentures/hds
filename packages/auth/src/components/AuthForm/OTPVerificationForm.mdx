import { Canvas, Meta, Story, ArgsTable } from '@storybook/addon-docs';

import StorybookHeader from '../../../../../components/StorybookHeader';

<Meta title="Components/Auth Form/OTP Login" />

<StorybookHeader
  title="HDS Authentication"
  command="npm install @highoutput/hds-auth"
/>

<Canvas>
  <Story id="components-auth-form-otp-login--default"></Story>
</Canvas>

<ArgsTable />
