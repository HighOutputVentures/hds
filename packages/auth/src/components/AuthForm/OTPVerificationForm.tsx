import { Box, Button } from '@chakra-ui/react';
import { TextField } from '@highoutput/hds-forms';
import { yupResolver } from '@hookform/resolvers/yup';
import React from 'react';
import { useForm } from 'react-hook-form';
import OTPForm, { OTPFormProps } from './OTPForm';
import {
  AuthenticateSchemaValues,
  generateEmailOTPSchema,
  GenerateEmailOTPSchemaValues,
} from './validation';

export interface OTPVerificationProps extends OTPFormProps {
  onSubmitEmailValue?(value: GenerateEmailOTPSchemaValues): void;
  onSubmitOTPValue?(value: AuthenticateSchemaValues): void;
  otpReceived: boolean;
  numberOfFields?: 4 | 6;
  otpType?: 'number' | 'alphanumeric';
  __testId?: string;
}

const OTPVerificationForm = (props: OTPVerificationProps) => {
  const {
    otpReceived,
    onSubmitEmailValue,
    numberOfFields,
    title,
    subTitle,
    onSubmitOTPValue,
    __testId,
  } = props;

  const { register, handleSubmit, formState } =
    useForm<GenerateEmailOTPSchemaValues>({
      resolver: yupResolver(generateEmailOTPSchema),
      defaultValues: {
        emailAddress: '',
      },
      shouldUnregister: true,
    });

  const onSubmitEmail = async (value: GenerateEmailOTPSchemaValues) => {
    if (onSubmitEmailValue) {
      onSubmitEmailValue(value);
    }
  };
  const onSubmitOTP = async (value: AuthenticateSchemaValues) => {
    if (onSubmitOTPValue) {
      onSubmitOTPValue(value);
    }
  };

  const { errors, isSubmitting } = formState;

  return (
    <Box>
      {!otpReceived ? (
        <Box
          as={'form'}
          data-testid="box.emailform.form"
          w={350}
          onSubmit={handleSubmit(onSubmitEmail)}
        >
          <TextField
            id="emailAddress"
            error={errors.emailAddress?.message}
            isDisabled={isSubmitting}
            placeholder={'Enter your email address'}
            __fieldTestId={__testId ?? 'hds.otp.verification.input.email'}
            {...register('emailAddress')}
          />

          <Button
            variant={'primary'}
            type="submit"
            isLoading={isSubmitting}
            data-testid="hds.otp.verification.submit.button"
            width={'100%'}
            marginTop={'10px'}
          >
            Sign In
          </Button>
        </Box>
      ) : (
        <Box data-testid="otp.component">
          <OTPForm
            title={title}
            subTitle={subTitle}
            onSubmitOTPValue={(value) => onSubmitOTP(value)}
            numberOfFields={numberOfFields}
          />
        </Box>
      )}
    </Box>
  );
};

export default OTPVerificationForm;
