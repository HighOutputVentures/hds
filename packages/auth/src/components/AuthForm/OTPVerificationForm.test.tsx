import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import OTPVerificationForm from './OTPVerificationForm';

describe('Email Login Component', () => {
  beforeEach(() => {
    render(<OTPVerificationForm otpReceived={false} />);
  });

  it('should render button submit for email', async () => {
    const submitButton = await screen.findAllByTestId(
      'hds.otp.verification.submit.button'
    );
    expect(submitButton).toHaveLength(1);
  });
  test('user clicks submit with no value or invalid input and renders error messages', async () => {
    const submit = await screen.findByTestId('box.emailform.form');
    await fireEvent.submit(submit);
    const errorFormControl = await screen.findAllByTestId(
      /hds.form.control.error.message/i
    );
    expect(errorFormControl).toHaveLength(1);
  });
});
describe('One Time Password Verification Component', () => {
  beforeEach(() => {
    render(<OTPVerificationForm otpReceived={true} />);
  });

  it('should render otp component', async () => {
    const otpComponent = await screen.findAllByTestId('otp.component');
    expect(otpComponent).toHaveLength(1);
  });
});

describe('Snapshot', () => {
  it('(OTPVerificationForm) Should match snapshot', () => {
    const component = renderer.create(<OTPVerificationForm otpReceived />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    component.unmount();
  });
});
