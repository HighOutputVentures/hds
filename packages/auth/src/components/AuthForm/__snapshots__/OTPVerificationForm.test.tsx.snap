// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot (OTPVerificationForm) Should match snapshot 1`] = `
<div
  className="css-0"
>
  <div
    className="css-0"
    data-testid="otp.component"
  >
    <form
      className="css-1ivgbss"
      data-testid="box.otpform.form"
      onSubmit={[Function]}
    >
      <div
        className="css-jmbq56"
      >
        <h1
          className="chakra-heading css-wlea3r"
        >
          Check your inbox
        </h1>
        <p
          className="chakra-text css-0"
        >
          We have sent a 6-digit code to your email 
        </p>
      </div>
      <div
        aria-label="Form Group"
        className="chakra-form-control css-0"
        data-testid="hds.form.control"
        role="group"
      >
        <div
          className="chakra-stack css-1tsaady"
        >
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
          <div
            className="css-iux5u6"
          >
            -
          </div>
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autoComplete="one-time-code"
            className="chakra-pin-input css-25hifv"
            data-testid="hds.pininput.pin"
            disabled={false}
            id="otp--1"
            inputMode="numeric"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder=""
            type="tel"
            value=""
          />
        </div>
      </div>
      <button
        className="chakra-button css-yrcen6"
        data-testid="hds.otp.submit.button"
        disabled={false}
        type="submit"
      >
        Sign In
      </button>
    </form>
  </div>
</div>
`;
